Add PEPS initialization based on fidelity maximization

pbrehmer opened the pull request:
This PR adds tools to initialize a PEPS based on maximizing the fidelity with respect to another PEPS. As we discussed on Tuesday, this would be useful e.g. to extract a single-site approximation of a SU evolved PEPS on a $2\times2$ unit cell. This would be great to have since SU/FU evolved initial PEPS can then be used for single-site variational optimization.

Note that I use TensorKit's embed branch in order to use the `embed!` function:
```
] add TensorKit#ld-embed
```

pbrehmer committed:
Add `maximize_fidelity!` (not yet working)

pbrehmer committed:
Fix environment initialization for differently sized PEPS and add docs

pbrehmer committed:
Add `single_site_fidelity_intiailize`

pbrehmer committed:
Replace network_value with simple norm

lkdvos commented on src/algorithms/time_evolution/fidelity_maximization.jl:
```diff
@@ -0,0 +1,134 @@
+end
+
+# maximal virtual space over unit cell
+_spacemax(peps::InfinitePEPS) = argmax(dim, map(p -> argmax(dim, domain(p)), unitcell(peps)))
```
I think you want to use `reduce(supremum, spaces)` to determine the maximal space, since this might be from a different site per sector.

pbrehmer replied:
I gave this a try now and I think it does the right thing, see the test

lkdvos commented on src/algorithms/time_evolution/fidelity_maximization.jl:
```diff
@@ -0,0 +1,134 @@
+is used on the environment bonds and kept fixed. The CTMRG contraction algorithm is specified
+via the `boundary_alg` `NamedTuple`.
+"""
+function maximize_fidelity!(
```
We could also just use `MPSKit.approximate!` here

pbrehmer replied:
Yes, I like that better :-)

I'll try to finish the rest over the next few days, but time is limited at the moment since I'm at a summer school so probably it's still gonna take a bit

lkdvos reviewed (commented):
I'll try and go over this in a bit more detail later, just wanted to leave these small comments before I forget

pbrehmer committed:
Merge branch 'master' into pb/fidelity-initialization

pbrehmer committed:
Change filelocation and rename to MPSKit.approximate!

pbrehmer committed:
Set up small test

pbrehmer committed:
Fix formatting

pbrehmer committed:
Rewrite `approximate!` and add non-uniform unit cell test

pbrehmer committed:
Merge branch 'master' into pb/fidelity-initialization

pbrehmer committed:
Fix _spacemax with reduce and add test

pbrehmer committed:
Remove +spacemax from test

pbrehmer committed:
Fix docs reference

pbrehmer committed:
Merge branch 'master' into pb/fidelity-initialization

leburgel committed:
Bump TensorKit compat

leburgel commented on src/algorithms/optimization/fidelity_initialize.jl:
```diff
@@ -0,0 +1,145 @@
+    @assert size(pepsdst) == size(pepssrc) "incompatible unit cell sizes"
+
+    # absorb src PEPS tensors into dst tensors in-place
+    for (pdst, psrc) in zip(unitcell(pepsdst), unitcell(pepssrc))
+        absorb!(pdst, psrc)
+    end
```
I think it would be better to have this in `single_site_fidelity_initialize` instead of inside the actual `approximate!` implementation. Now we're really throwing away the initial guess, which is not really what we want if we start from a very good guess already for example.

leburgel replied:
My concern wasn't so much the fact that it was in-place, but rather that doing the absorption step by default pretty much discards the initial guess. I would prefer to be able to actually use a really good initial guess for `pepsdst`.

So definitely agree with having a  non-mutating `approximate`, but even there much of the initial guess is thrown away by the `absorb` step. Similar to the MPSKit algorithms, it seems better to really start from the initial guess that was given as is, and keep the step that prepares a good initial guess for the routine calling `approximate`, i.e., `single_site_fidelity_initialize`.

pbrehmer committed:
Add `approximate` method

pbrehmer committed:
Add `LoggingExtras.withlevel` logging

pbrehmer committed:
Update tests

pbrehmer commented on test/toolbox/fidelity_initialize.jl:
```diff
@@ -0,0 +1,34 @@
+using Test
+
+@testset "Fidelity initialization with approximate!" begin
+    Random.seed!(18092025)
```
Not really, the reason I fixed the seed was that the CTMRG contractions will not converge in the beginning of the fidelity maximization because it seems that the $3\times3$ unit cell and random spaces makes this hard to contract. Fixing the seed doesn't really change this but I wanted to at least make it deterministic.

As a general question: Do we have a feeling on how well this fidelity approximation should be able to converge? Because for the case where we approximate the $3\times3$ PEPS with a single-site PEPS, the infidelity will not drop below 0.1 but it does plateau somewhat.

leburgel replied:
I see, then I guess using a specific seed is fine for now until we have some tools to properly address #255, which should then also remove the need to finetune here.

I don't think this simple power method here should converge very well, as long as it gets in the right ballpark it should be enough for an initialization. From the point of view of proper fidelity maximization, the gradient used as computed by `_∂local_norm` is really only the partial gradient of the fidelity contraction with respect to the PEPS tensor being optimized over. To get the full gradient we would have to take into account the contributions that come from the environment tensors in the fidelity contraction. We could do this by backpropagating through the entire contraction step just like we do for energy minimizations, but that's probably overkill here. It might be good to verify that using the full gradient gives a clean maximization of the fidelity, but that would be more expensive and likely unnecessary when just getting a good initial state.

If this approach doesn't solve the issues then we can try using the full fidelity gradient instead.

pbrehmer commented on src/algorithms/optimization/fidelity_initialize.jl:
```diff
@@ -0,0 +1,145 @@
+"""
+function MPSKit.approximate!(
+        pepsdst::InfinitePEPS, pepssrc::InfinitePEPS, envspace;
+        maxiter = 10, tol = 1.0e-3, boundary_alg = (; verbosity = 1)
```
Good point, added it now

pbrehmer committed:
Actually return single-site PEPS

pbrehmer committed:
Add missing break

pbrehmer committed:
Merge branch 'master' into pb/fidelity-initialization

pbrehmer committed:
Remove export of `single_site_fidelity_initialize`

pbrehmer committed:
Remove `absorb!` from `approximate!`

pbrehmer committed:
Add comments

pbrehmer committed:
Merge branch 'pb/fidelity-initialization' of github.com:quantumghent/PEPSKit.jl into pb/fidelity-initialization

pbrehmer committed:
Merge branch 'master' into pb/fidelity-initialization

pbrehmer committed:
Add `ApproximateAlgorithm` abstract type and `FidelityMaxCrude` algorithm struct

pbrehmer committed:
Match `approximate!` function signature with MPSKit

pbrehmer committed:
Remove `approximate!` and fix errors

pbrehmer committed:
Add small `approximate` test

pbrehmer committed:
Merge branch 'pb/fidelity-initialization' of github.com:quantumghent/PEPSKit.jl into pb/fidelity-initialization

pbrehmer committed:
Merge branch 'master' into pb/fidelity-initialization

pbrehmer committed:
Merge branch 'master' into pb/fidelity-initialization

pbrehmer committed:
Merge branch 'master' into pb/fidelity-initialization

pbrehmer commented:
@leburgel @Yue-Zhengyuan Quick update on running SU with staggered charges on the physical spaces (as discussed yesterday): This does seem to work indeed and I obtain SU energies which are quite similar to the previous runs. Note that I first had to fix the index rotation functions to make this work, see #267. I haven't run variational optimizations on these states yet, but I'll keep you posted.

leburgel commented:
@pbrehmer, how would you feel about splitting off and adding the `approximate` part of this at least? I think the single-site replacement was a bit tricky conceptually, but the crude fidelity optimization by itself could be a very useful tool, so I think it might be good to revive that part.

pbrehmer commented:
> @pbrehmer, how would you feel about splitting off and adding the `approximate` part of this at least? I think the single-site replacement was a bit tricky conceptually, but the crude fidelity optimization by itself could be a very useful tool, so I think it might be good to revive that part.

I agree, the crude fidelity `approximate` would indeed be helpful. Then I will remove the single-site initialization and just leave the `approximate` part as is.

pbrehmer committed:
Remove single-site fidelity initialization

pbrehmer committed:
Merge branch 'master' into pb/fidelity-initialization

pbrehmer committed:
Merge branch 'pb/fidelity-initialization' of github.com:quantumghent/PEPSKit.jl into pb/fidelity-initialization

pbrehmer committed:
Fix `approximate` test

pbrehmer commented:
I removed the single-site initialization and fixed the test. Of course one could add more test cases but I think this is also fine as is. Would be good to go for me.

Yue-Zhengyuan commented on src/algorithms/approximate.jl:
```diff
@@ -0,0 +1,211 @@
+    log = MPSKit.IterLog("Approximate")
+    return LoggingExtras.withlevel(; alg.verbosity) do
+        # normalize reference PEPS
+        peps_init = ψ # smaller bond spaces
```
I think ψ has larger bond spaces?

pbrehmer replied:
I think in the original context where I implemented this we were treating the case where we wanted to initialize a PEPS with larger $D$ from a smaller $D$ PEPS. (I guess one frequently encounters this situation when optimizing PEPS where larger $D$ optimization runs are hard to initialize properly.) But in principle `ψ` could of course also have a larger bond space - this would correspond to a lower rank approximation of a large $D$ PEPS. In principle both cases should be allowed (I think?), so I'll just remove the comment.

Yue-Zhengyuan commented on src/algorithms/approximate.jl:
```diff
@@ -0,0 +1,211 @@
+$(TYPEDEF)
+
+PEPS approximation algorithm maximizing the fidelity by successively applying the fidelity
+derivative with respect to the approximator PEPS.
```
Can we add a more detailed description of the iterative process?

pbrehmer replied:
I added some text describing the algorithm.

Yue-Zhengyuan commented on test/toolbox/approximate.jl:
```diff
@@ -0,0 +1,30 @@
+        maxiter = maxiter_approx,
+        boundary_alg = (; maxiter = maxiter_boundary, trunc = truncrank(6), verbosity = 1)
+    )
+    Random.seed!(sd) # reseeding with the same seed is required for the test to pass
```
Is this test merely testing if the algorithm selection works, which is why the two runs need exactly the same `peps1` and `peps2`?

pbrehmer replied:
Yes indeed. It doesn't hurt testing that since the test itself is quite cheap.

Yue-Zhengyuan commented on src/algorithms/approximate.jl:
```diff
@@ -0,0 +1,211 @@
+        peps′ = _∂local_norm(peps_init, envnw)
+        for iter in 1:(alg.maxiter)
+            # compute fidelity from ∂norm
+            fid = abs2(_local_norm(peps, peps′))
```
I suggest we describe in the docstring how the "crude fidelity" we are using is defined, and how it is related to the exact `⟨ψ₀|ψ⟩`.

pbrehmer replied:
I added some words on this in the algorithm description.

Yue-Zhengyuan commented on src/algorithms/approximate.jl:
```diff
@@ -0,0 +1,211 @@
+        peps_init = ψ # smaller bond spaces
+        envspace = domain(ψ₀[1])[1]
+        env₀, = leading_boundary(CTMRGEnv(peps_init, envspace), peps_init, alg.boundary_alg)
+        peps_init /= sqrt(abs(_local_norm(peps_init, peps_init, env₀))) # normalize to ensure that fidelity is bounded by 1
```
In Julia, doing `ψ /= nrm` is exactly equivalent to `ψ = ψ / nrm` (@lkdvos correct me if I'm wrong, or the behavior is defined differently for PEPS/TensorMaps). After this assignment (rebinding), any changes to `ψ` does not affect the input variable. Then the function is not mutating, and we can avoid introducing new names like `peps_init`, `peps`.

Also, `env₀` being the environment for `ψ`, and `env` being the environment for `ψ₀` looks off to me.

pbrehmer replied:
I introduced new names on purpose since I felt that was a bit cleaner and made it easier to read to iterative process. Note that `env₀` is the environment of `peps₀` (see updated name) and `env` of `peps`, i.e. the `₀` index refers to the zeroth iteration step of the `approximate` routine, not the initial guess. Instead, the first iteration step is determined by the initial guess `ψ₀` since a perfect initial guess would immediately terminate the iteration. This is all slightly confusing so I wanted to work with different names.

pbrehmer committed:
Merge remote-tracking branch 'origin/master' into pb/fidelity-initialization

pbrehmer committed:
Apply suggestions

Yue-Zhengyuan commented on test/toolbox/approximate.jl:
What do you think about adding a more elementary test?
```julia
using Test
using Random
using TensorKit
using PEPSKit
using PEPSKit: _∂local_norm, _local_norm

sd = 123456
Random.seed!(sd)
Vphys, Vbond, Venv = ℂ^2, ℂ^3, ℂ^6

ket = InfinitePEPS(Vphys, Vbond)
bra = InfinitePEPS(Vphys, Vbond)
network = InfiniteSquareNetwork(ket, bra)
env = CTMRGEnv(network, Venv)

nrm1 = _local_norm(ket, bra, env)
∂norm = _∂local_norm(ket, env)
nrm2 = _local_norm(bra, ∂norm)
@test nrm1 ≈ nrm2
```
It helps me (and maybe others) understand what the underlying functions are doing.

pbrehmer replied:
Sure, I really do like that this illustrates how the internal functions are used.

Yue-Zhengyuan commented on src/algorithms/approximate.jl:
```diff
@@ -0,0 +1,218 @@
+derivative with respect to the approximator PEPS.
+
+Starting from an initial guess, the algorithm first computes the CTMRG environment of the ⟨ψᵢ|ψᵢ₊₁⟩
+network. From that, the fidelity derivative ∂F/∂ψᵢ₊₁ is computed for each PEPS tensor in the unit cell,
```
Please explain what is `F` before introducing the notation.

Yue-Zhengyuan commented on src/algorithms/approximate.jl:
```diff
@@ -0,0 +1,218 @@
+Starting from an initial guess, the algorithm first computes the CTMRG environment of the ⟨ψᵢ|ψᵢ₊₁⟩
+network. From that, the fidelity derivative ∂F/∂ψᵢ₊₁ is computed for each PEPS tensor in the unit cell,
+corresponding to the norm network of ψᵢ where the respective bra PEPS tensor is missing. Then we update
+ψᵢ = ψᵢ₊₁ and ψᵢ₊₁ = ∂F/∂ψᵢ₊₁, and compute the fidelity |⟨ψᵢ|ψᵢ₊₁⟩|². This is iterated until the
```
When describing an iterative process I would write something like `ψᵢ₊₁ = f(ψᵢ)`. So `ψᵢ = ψᵢ₊₁` feels a bit unnatural. Again, please explain what is `F`.

pbrehmer replied:
I now omit the use of `F` (which was technically wrong anyway).

Yue-Zhengyuan commented on src/algorithms/approximate.jl:
```diff
@@ -0,0 +1,218 @@
+end
+
+# contract CTMRG environment leaving open the bra-PEPS virtual and physical bonds
+function _∂contract_site(ind::Tuple{Int, Int}, peps::InfinitePEPS, env::CTMRGEnv)
```
For clarity, maybe we just rename `peps` to `ket`?

Yue-Zhengyuan commented on src/algorithms/approximate.jl:
```diff
@@ -0,0 +1,218 @@
+
+Compute the `InfinitePEPS` resulting from removing the bra PEPS tensors in `_local_norm`.
+"""
+function _∂local_norm(peps::InfinitePEPS, env::CTMRGEnv)
```
How about renaming `peps` to `ket`?

Yue-Zhengyuan reviewed (commented):
Some more nitpicky comments

Yue-Zhengyuan commented on test/toolbox/approximate.jl:
```diff
@@ -0,0 +1,31 @@
+        )
+    )
+
+    @test peps_approx1 ≈ peps_approx2
```
As a less trivial test, can we do two approximations with two different seeds (leading to different CTMRGEnv during the run), and finally check if the fidelity of the two results is close enough to 1?

pbrehmer replied:
Yes, I'll add that to the existing test set.

Yue-Zhengyuan replied:
On my Mac this new test fails miserably...

pbrehmer replied:
How miserably does it fail? In general I think we can't expect the fidelities to approach 1 very closely since the algorithm is rather crude but I thought that being within 10% should be possible. Does is get better for other seeds?

pbrehmer replied:
Okay, trying around with different seeds this test will always fail. Maybe the tolerance is too strict: If one evaluates the fidelity of two random PEPS, it will in general be very small. So perhaps it is enough if the fidelity is already of the order of $10^{-1}$ (as opposed to $10^{-7}$ for two random PEPS).

Yue-Zhengyuan replied:
For me the fidelity between two runs with different seeds can be ad low as 0.3. If random states are hard, how about using SU to get a physical state and try to approximate it?

pbrehmer replied:
I don't think the problem is the initial state, it's rather the algorithm itself which will only bring two states in the same ballpark. One can already see in the convergence itself that it is hard to converge the fidelity below an error of $10^{-2}$. But it would be interesting to see what happens if one initializes with SU.

Yue-Zhengyuan commented on src/algorithms/approximate.jl:
```diff
@@ -0,0 +1,218 @@
+function _local_norm(ket::InfinitePEPS, bra::InfinitePEPS, env::CTMRGEnv)
+    return sum(ind -> contract_local_norm((ind,), ket, bra, env), eachcoordinate(ket))
+end
+function _local_norm(peps::InfinitePEPS, ∂norm::InfinitePEPS)
```
Also, I suggest renaming `peps` to `bra` here. Maybe also a more descriptive name for `∂norm` to emphasize that it is the derivative *with respect to the bra tensors*?

pbrehmer replied:
I find that `∂norm` is fine as a name, especially since it does have a docstring where it is explicitly stated that the derivative is with respect to the bra.

pbrehmer committed:
Apply some more suggestions

pbrehmer committed:
Fix lil` approximate test error

pbrehmer committed:
Loosen tolerance on fidelity test

pbrehmer committed:
Lower approx tolerance even more

Yue-Zhengyuan commented:
If I take a random PEPS `ψ1`, and obtain another `ψ2` by performing a gauge transformation (inserting `X inv(X)` on virtual bonds), will the crude fidelity defined here evaluates to 1? This is not obvious to me by looking at the code.

Yue-Zhengyuan commented on test/toolbox/approximate.jl:
```diff
@@ -57,5 +57,5 @@ end
     ∂norm = _∂local_norm(peps_approx1, envnw)
     fid = abs2(_local_norm(peps_approx3, ∂norm))
-    @test fid ≈ 1 atol = 0.8 # can't expect much more than 10⁻¹
+    @test fid ≈ 1 atol = 0.9 # judging from different seeds, one generally can't expect much more than 10⁻¹ fidelity
```
I think we just test `fid > 0.1` if it is so bad

pbrehmer committed:
Rephrase test without `isapprox`

pbrehmer commented:
> If I take a random PEPS `ψ1`, and obtain another `ψ2` by performing a gauge transformation (inserting `X inv(X)` on virtual bonds), will the crude fidelity defined here evaluates to 1? This is not obvious to me by looking at the code.

I don't think it will evaluate to 1 due to the way CTMRG (without spatial symmetries) is not gauge-invariant if you insert a gauge transformation on the virtual bonds since the north and east edge will absorb $X$ whereas the south and west edge will absorb $X^{-1}$. It would only be one if that inner product contraction environment of `InfiniteSquareNetwork(ψ1, ψ2)` would produce the same network value as the respective `ψ1` and `ψ2` norm networks - which it generally won't due to the gauge transformation, I believe.

Edit: If that is true, then that also kind of answers why the fidelity test won't approach 1, since two different initializations won't end up in the same gauge.

Yue-Zhengyuan commented:
Then I suggest we also explain in docstring how the fidelity is crudely approximated so that users can have the right expectations.

Yue-Zhengyuan commented:
Just remembered this discussion #316 that is very relevant.

pbrehmer committed:
Add a paragraph on crudeness of the algorithm

Yue-Zhengyuan commented:
~I think that's all I wanted to say.~ @leburgel do you have additional comments?

Yue-Zhengyuan commented:
Oh, can we maybe use BP gauging fixing during the iteration to improve the robustness?

pbrehmer commented:
> Oh, can we maybe use BP gauging fixing during the iteration to improve the robustness?

I gave this a quick try. I tried both gauging before the power-method iteration and during the iteration. This didn't really help to stabilize the algorithm from what I found. In case you had something more specific in mind, let me know. I would rather be in favor of leaving the algorithm as is because I think it already is useful for the original use case we had in mind (optimization initialization). Maybe we can try to come up with a more refined algorithm in the future but I also don't want to spend much more time on this PR right now.
